From 1f4374af204a516b85497c41e815503ae443b88d Mon Sep 17 00:00:00 2001
From: Martin KaFai Lau <kafai@fb.com>
Date: Wed, 24 Mar 2021 18:52:14 -0700
Subject: [PATCH 338/921] [Backport] libbpf: Refactor codes for finding btf id
 of a kernel symbol

mainline inclusion
from mainline-v5.13-rc1
commit 774e132e83d0f10a7ebbfe7db1debdaed6013f83
category: feature
bugzilla: N/A

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=774e132e83d0f10a7ebbfe7db1debdaed6013f83

------------------------------------------------------

This patch refactors code, that finds kernel btf_id by kind
and symbol name, to a new function find_ksym_btf_id().

It also adds a new helper __btf_kind_str() to return
a string by the numeric kind value.

Signed-off-by: Martin KaFai Lau <kafai@fb.com>
Signed-off-by: Alexei Starovoitov <ast@kernel.org>
Acked-by: Andrii Nakryiko <andrii@kernel.org>
Link: https://lore.kernel.org/bpf/20210325015214.1547069-1-kafai@fb.com

Conflicts:
	tools/lib/bpf/libbpf.c

Signed-off-by: Yang Jihong <yangjihong1@huawei.com>
---
 tools/lib/bpf/libbpf.c | 39 +++++++++++++++++++++++++++++----------
 1 file changed, 29 insertions(+), 10 deletions(-)

diff --git a/tools/lib/bpf/libbpf.c b/tools/lib/bpf/libbpf.c
index 63b3fd6ffe75..b4e1f98f52d2 100644
--- a/tools/lib/bpf/libbpf.c
+++ b/tools/lib/bpf/libbpf.c
@@ -1898,9 +1898,9 @@ resolve_func_ptr(const struct btf *btf, __u32 id, __u32 *res_id)
 	return btf_is_func_proto(t) ? t : NULL;
 }
 
-static const char *btf_kind_str(const struct btf_type *t)
+static const char *__btf_kind_str(__u16 kind)
 {
-	switch (btf_kind(t)) {
+	switch (kind) {
 	case BTF_KIND_UNKN: return "void";
 	case BTF_KIND_INT: return "int";
 	case BTF_KIND_PTR: return "ptr";
@@ -1921,6 +1921,11 @@ static const char *btf_kind_str(const struct btf_type *t)
 	}
 }
 
+static const char *btf_kind_str(const struct btf_type *t)
+{
+	return __btf_kind_str(btf_kind(t));
+}
+
 /*
  * Fetch integer attribute of BTF map definition. Such attributes are
  * represented using a pointer to an array, in which dimensionality of array
@@ -7169,6 +7174,24 @@ static int bpf_object__read_kallsyms_file(struct bpf_object *obj)
 	return err;
 }
 
+static int find_ksym_btf_id(struct bpf_object *obj, const char *ksym_name,
+			    __u16 kind, struct btf **res_btf)
+{
+	int id;
+	struct btf *btf;
+
+	btf = obj->btf_vmlinux;
+	id = btf__find_by_name_kind(btf, ksym_name, kind);
+	if (id <= 0) {
+		pr_warn("extern (%s ksym) '%s': failed to find BTF ID in kernel BTF(s).\n",
+			__btf_kind_str(kind), ksym_name);
+		return -ESRCH;
+	}
+
+	*res_btf = btf;
+	return id;
+}
+
 static int bpf_object__resolve_ksym_var_btf_id(struct bpf_object *obj,
 					       struct extern_desc *ext)
 {
@@ -7176,15 +7199,11 @@ static int bpf_object__resolve_ksym_var_btf_id(struct bpf_object *obj,
 	__u32 targ_type_id, local_type_id;
 	const char *targ_var_name;
 	int id, err;
-	struct btf *btf;
+	struct btf *btf = NULL;
 
-	btf = obj->btf_vmlinux;
-	id = btf__find_by_name_kind(btf, ext->name, BTF_KIND_VAR);
-	if (id <= 0) {
-		pr_warn("extern (ksym) '%s': failed to find BTF ID in vmlinux BTF.\n",
-			ext->name);
-		return -ESRCH;
-	}
+	id = find_ksym_btf_id(obj, ext->name, BTF_KIND_VAR, &btf);
+	if (id < 0)
+		return id;
 
 	/* find local type_id */
 	local_type_id = ext->ksym.type_id;
-- 
2.17.1

